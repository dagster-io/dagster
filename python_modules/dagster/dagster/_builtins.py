import typing

Any = typing.Any
Bool = bool
Float = float
Int = int
Nothing: type[None] = type(None)
String = str


class BuiltinEnum:
    ANY = Any
    BOOL = Bool
    FLOAT = Float
    INT = Int
    NOTHING = Nothing
    STRING = String

    @classmethod
    def contains(cls, value: object) -> bool:
        for ttype in [cls.ANY, cls.BOOL, cls.FLOAT, cls.INT, cls.STRING, cls.NOTHING]:
            if value == ttype:
                return True

        return False
