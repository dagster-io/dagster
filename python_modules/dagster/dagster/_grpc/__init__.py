"""Tools for accessing core Dagster APIs over a GRPC mechanism.

GRPC is intended to be used in all cases where host processes communicate with user processes, both
locally (over UDS on MacOS and Unix, and over a local port on Windows) and when communicating with
remote Dagster user proceses (e.g., containers).

The GRPC layer is not intended to supplant the dagster-graphql layer, which should still be used to
drive web frontends like the Dagster UI.
"""

# Prevents dagster._core => dagster._grpc.types => dagster._api => dagster._grpc.types circular imports
import dagster._core.remote_representation.external_data  # noqa

from dagster._grpc.types import (
    CanCancelExecutionRequest as CanCancelExecutionRequest,
    CanCancelExecutionResult as CanCancelExecutionResult,
    CancelExecutionRequest as CancelExecutionRequest,
    CancelExecutionResult as CancelExecutionResult,
    ExecuteExternalJobArgs as ExecuteExternalJobArgs,
    ExecuteRunArgs as ExecuteRunArgs,
    ExecuteStepArgs as ExecuteStepArgs,
    ExecutionPlanSnapshotArgs as ExecutionPlanSnapshotArgs,
    ExternalJobArgs as ExternalJobArgs,
    ExternalScheduleExecutionArgs as ExternalScheduleExecutionArgs,
    GetCurrentImageResult as GetCurrentImageResult,
    JobSubsetSnapshotArgs as JobSubsetSnapshotArgs,
    ListRepositoriesInput as ListRepositoriesInput,
    ListRepositoriesResponse as ListRepositoriesResponse,
    LoadableRepositorySymbol as LoadableRepositorySymbol,
    NotebookPathArgs as NotebookPathArgs,
    PartitionArgs as PartitionArgs,
    PartitionNamesArgs as PartitionNamesArgs,
    PartitionSetExecutionParamArgs as PartitionSetExecutionParamArgs,
    ResumeRunArgs as ResumeRunArgs,
    SensorExecutionArgs as SensorExecutionArgs,
    ShutdownServerResult as ShutdownServerResult,
    StartRunResult as StartRunResult,
)
from dagster._grpc.utils import get_loadable_targets as get_loadable_targets
