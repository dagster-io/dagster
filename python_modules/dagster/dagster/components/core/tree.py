import importlib
from functools import cached_property
from pathlib import Path
from types import ModuleType

from dagster_shared.record import record
from typing_extensions import TypeVar

from dagster._core.definitions.definitions_class import Definitions
from dagster.components.component.component import Component
from dagster.components.core.context import ComponentLoadContext, use_component_load_context
from dagster.components.core.defs_module import DefsFolderComponent
from dagster.components.core.load_defs import get_library_json_enriched_defs

PLUGIN_COMPONENT_TYPES_JSON_METADATA_KEY = "plugin_component_types_json"

TComponent = TypeVar("TComponent", bound=Component)


@record(
    checked=False,  # cant handle ModuleType
)
class ComponentTree:
    defs_module: ModuleType
    project_root: Path

    @staticmethod
    def load(path_within_project: Path):
        """Using the provided path, find the nearest parent python project and load the
        ComponentTree using its configuration.
        """
        from dagster_dg_core.context import DgContext

        # replace with dagster_shared impl of path crawl and config resolution
        dg_context = DgContext.for_project_environment(path_within_project, command_line_config={})

        return ComponentTree(
            defs_module=importlib.import_module(dg_context.defs_module_name),
            project_root=dg_context.root_path,
        )

    @cached_property
    def load_context(self):
        return ComponentLoadContext.for_module(
            defs_module=self.defs_module,
            project_root=self.project_root,
        )

    @cached_property
    def root(self) -> DefsFolderComponent:
        return DefsFolderComponent.get(self.load_context)

    def load_defs(self) -> Definitions:
        with use_component_load_context(self.load_context):
            return Definitions.merge(
                self.root.build_defs(self.load_context),
                get_library_json_enriched_defs(),
            )

    def load_defs_at_path(self, defs_path: Path) -> Definitions:
        # impl to be fleshed out to flexibly handle different path types (str, list[str], ...)
        for cp, c in self.root.iterate_path_component_pairs():
            if cp.file_path.relative_to(self.root.path) == defs_path:
                ctx = self.load_context.for_path(cp.file_path)
                with use_component_load_context(ctx):
                    return c.build_defs(ctx)

        raise Exception(f"No component found for path {defs_path}")

    def get_all_components(
        self,
        of_type: type[TComponent],
    ) -> list[TComponent]:
        """Get all components from this context that are instance of the specified type."""
        return [
            component
            for component in self.root.iterate_components()
            if isinstance(component, of_type)
        ]
