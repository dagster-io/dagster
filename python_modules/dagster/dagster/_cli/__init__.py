import click

from dagster._cli.api import api_cli
from dagster._cli.asset import asset_cli
from dagster._cli.code_server import code_server_cli
from dagster._cli.debug import debug_cli
from dagster._cli.definitions import definitions_cli
from dagster._cli.dev import dev_command
from dagster._cli.instance import instance_cli
from dagster._cli.job import job_cli
from dagster._cli.json import json_cli
from dagster._cli.project import project_cli
from dagster._cli.run import run_cli
from dagster._cli.schedule import schedule_cli
from dagster._cli.sensor import sensor_cli
from dagster.version import __version__


def create_dagster_cli():
    commands = {
        "api": api_cli,
        "job": job_cli,
        "run": run_cli,
        "instance": instance_cli,
        "schedule": schedule_cli,
        "sensor": sensor_cli,
        "asset": asset_cli,
        "debug": debug_cli,
        "project": project_cli,
        "dev": dev_command,
        "code-server": code_server_cli,
        "definitions": definitions_cli,
        "json": json_cli,
    }

    @click.group(
        commands=commands,
        context_settings={"max_content_width": 120, "help_option_names": ["-h", "--help"]},
    )
    @click.version_option(__version__, "--version", "-v")
    def group():
        """CLI tools for working with Dagster."""

    return group


ENV_PREFIX = "DAGSTER_CLI"
cli = create_dagster_cli()


def main():
    cli(auto_envvar_prefix=ENV_PREFIX)
