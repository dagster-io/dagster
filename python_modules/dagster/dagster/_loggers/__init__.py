import logging
from collections.abc import Mapping, Sequence
from typing import TYPE_CHECKING, Optional, Tuple  # noqa: F401, UP035

import coloredlogs

from dagster import _seven
from dagster._config import Field
from dagster._core.definitions.logger_definition import LoggerDefinition, logger
from dagster._core.log_manager import (
    LOG_RECORD_EVENT_ATTR,
    LOG_RECORD_METADATA_ATTR,
    DagsterLogRecordMetadata,
)
from dagster._core.utils import coerce_valid_log_level
from dagster._serdes import pack_value
from dagster._utils.log import create_console_logger

if TYPE_CHECKING:
    from dagster._core.execution.context.logger import InitLoggerContext
    from dagster._core.instance import DagsterInstance


@logger(
    Field(
        {
            "log_level": Field(
                str,
                is_required=False,
                default_value="INFO",
                description="The logger's threshold.",
            ),
            "name": Field(
                str,
                is_required=False,
                default_value="dagster",
                description="The name of your logger.",
            ),
        },
        description="The default colored console logger.",
    ),
    description="The default colored console logger.",
)
def colored_console_logger(init_context: "InitLoggerContext") -> logging.Logger:
    """This logger provides support for sending Dagster logs to stdout in a colored format. It is
    included by default on jobs which do not otherwise specify loggers.
    """
    return create_console_logger(
        name=init_context.logger_config["name"],
        level=coerce_valid_log_level(init_context.logger_config["log_level"]),
    )


class JsonLogFormatter(logging.Formatter):
    def format(self, record):
        dict_to_dump = {}
        for k, v in record.__dict__.items():
            if k == LOG_RECORD_EVENT_ATTR:
                # Redundant with the "dagster_event" field under "dagster_meta"
                continue
            elif k == LOG_RECORD_METADATA_ATTR:
                # Events objects are not always JSON-serializable, so need to pack them first
                json_serializable_event = pack_value(v[LOG_RECORD_EVENT_ATTR])
                json_serializable_dagster_meta = DagsterLogRecordMetadata(
                    **{**v, "dagster_event": json_serializable_event}
                )
                dict_to_dump[LOG_RECORD_METADATA_ATTR] = json_serializable_dagster_meta
            else:
                dict_to_dump[k] = v

        return _seven.json.dumps(dict_to_dump)


@logger(
    Field(
        {
            "log_level": Field(
                str,
                is_required=False,
                default_value="INFO",
                description="The logger's threshold.",
            ),
            "name": Field(
                str,
                is_required=False,
                default_value="dagster",
                description="The name of your logger.",
            ),
        },
        description="A JSON-formatted console logger.",
    ),
    description="A JSON-formatted console logger.",
)
def json_console_logger(init_context: "InitLoggerContext") -> logging.Logger:
    """This logger provides support for sending Dagster logs to stdout in json format.

    Example:
        .. code-block:: python

            from dagster import op, job
            from dagster.loggers import json_console_logger

            @op
            def hello_op(context):
                context.log.info('Hello, world!')
                context.log.error('This is an error')

            @job(logger_defs={'json_logger': json_console_logger})])
            def json_logged_job():
                hello_op()

    """
    level = coerce_valid_log_level(init_context.logger_config["log_level"])
    name = init_context.logger_config["name"]

    klass = logging.getLoggerClass()
    logger_ = klass(name, level=level)

    handler = coloredlogs.StandardErrorHandler()

    handler.setFormatter(JsonLogFormatter())
    logger_.addHandler(handler)

    return logger_


def default_system_loggers(
    instance: Optional["DagsterInstance"],
) -> Sequence[tuple["LoggerDefinition", Mapping[str, object]]]:
    """If users don't provide configuration for any loggers, we instantiate these loggers with the
    default config.

    Returns:
        List[Tuple[LoggerDefinition, dict]]: Default loggers and their associated configs.
    """
    log_level = instance.python_log_level if (instance and instance.python_log_level) else "DEBUG"
    return [(colored_console_logger, {"name": "dagster", "log_level": log_level})]


def default_loggers() -> Mapping[str, "LoggerDefinition"]:
    return {"console": colored_console_logger}
