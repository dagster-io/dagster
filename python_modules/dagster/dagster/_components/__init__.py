from dagster_shared.serdes.objects import LibraryObjectKey as LibraryObjectKey

from dagster._components.component.component import Component as Component
from dagster._components.component.component_loader import component as component
from dagster._components.component.component_scaffolder import (
    DefaultComponentScaffolder as DefaultComponentScaffolder,
)
from dagster._components.component_scaffolding import scaffold_component as scaffold_component
from dagster._components.core.context import ComponentLoadContext as ComponentLoadContext
from dagster._components.core.load_defs import (
    build_component_defs as build_component_defs,
    load_defs as load_defs,
)
from dagster._components.resolved.base import Resolvable as Resolvable
from dagster._components.resolved.context import ResolutionContext as ResolutionContext
from dagster._components.resolved.core_models import (
    AssetAttributesModel as AssetAttributesModel,
    AssetPostProcessorModel as AssetPostProcessorModel,
    ResolvedAssetSpec as ResolvedAssetSpec,
)
from dagster._components.resolved.metadata import ResolvableFieldInfo as ResolvableFieldInfo
from dagster._components.resolved.model import (
    Injectable as Injectable,
    Injected as Injected,
    Model as Model,
    Resolver as Resolver,
)
from dagster._components.scaffold.scaffold import (
    Scaffolder as Scaffolder,
    ScaffolderUnavailableReason as ScaffolderUnavailableReason,
    ScaffoldRequest as ScaffoldRequest,
)
