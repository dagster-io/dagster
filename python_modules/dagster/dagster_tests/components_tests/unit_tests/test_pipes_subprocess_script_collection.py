import importlib
from pathlib import Path

import dagster as dg
from dagster.components.core.tree import LegacyAutoloadingComponentTree

LOCATION_PATH = Path(__file__).parent.parent / "code_locations" / "python_script_location"


def test_load_from_path(snapshot) -> None:
    module = importlib.import_module(
        "dagster_tests.components_tests.code_locations.python_script_location.defs"
    )
    tree = LegacyAutoloadingComponentTree.from_module(
        defs_module=module, project_root=Path(__file__).parent
    )
    snapshot.assert_match(tree.to_string_representation(include_load_and_build_status=True))
    tree.load_root_component()
    snapshot.assert_match(tree.to_string_representation(include_load_and_build_status=True))

    defs = tree.build_defs()
    assert defs.resolve_asset_graph().get_all_asset_keys() == {
        dg.AssetKey("a"),
        dg.AssetKey("b"),
        dg.AssetKey("c"),
        dg.AssetKey("up1"),
        dg.AssetKey("up2"),
        dg.AssetKey("override_key"),
        dg.AssetKey("cool_script"),
        dg.AssetKey("a_dash"),
        dg.AssetKey("b_dash"),
        dg.AssetKey("c_dash"),
        dg.AssetKey("up1_dash"),
        dg.AssetKey("up2_dash"),
        dg.AssetKey("override_key_dash"),
        dg.AssetKey("foo"),
        dg.AssetKey("bar"),
        dg.AssetKey("foo_def_py"),
        dg.AssetKey("bar_def_py"),
        dg.AssetKey("from_defs_one"),
        dg.AssetKey("from_defs_two"),
    }
    assert defs.component_tree

    snapshot.assert_match(tree.to_string_representation(include_load_and_build_status=True))


def test_load_from_location_path() -> None:
    module = importlib.import_module(
        "dagster_tests.components_tests.code_locations.python_script_location.defs.scripts"
    )
    defs = dg.load_defs(module, project_root=Path(__file__).parent)

    assert defs.resolve_asset_graph().get_all_asset_keys() == {
        dg.AssetKey("a"),
        dg.AssetKey("b"),
        dg.AssetKey("c"),
        dg.AssetKey("up1"),
        dg.AssetKey("up2"),
        dg.AssetKey("override_key"),
    }
