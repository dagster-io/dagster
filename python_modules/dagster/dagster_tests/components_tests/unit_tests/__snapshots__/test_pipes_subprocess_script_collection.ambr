# serializer version: 1
# name: test_load_from_path
  '''
  ├── multiple_component_instances
  │   └── component.py
  │       ├── bar (PythonScriptComponent)
  │       └── foo (PythonScriptComponent)
  ├── multiple_component_instances_defs_py
  │   └── component.py
  │       ├── bar (PythonScriptComponent)
  │       └── foo (PythonScriptComponent)
  ├── multiple_definitions_calls
  │   └── defs.py
  ├── script_python_decl
  │   ├── component.py
  │   │   └── load (PythonScriptComponent)
  │   └── cool_script.py
  ├── scripts
  │   ├── defs.yaml[0] (PythonScriptComponent)
  │   ├── defs.yaml[1] (PythonScriptComponent)
  │   └── defs.yaml[2] (PythonScriptComponent)
  └── triple_dash_scripts
      ├── defs.yaml[0] (PythonScriptComponent)
      ├── defs.yaml[1] (PythonScriptComponent)
      └── defs.yaml[2] (PythonScriptComponent)
  '''
# ---
# name: test_load_from_path.1
  '''
  ├── multiple_component_instances (loaded)
  │   └── component.py (loaded)
  │       ├── bar (PythonScriptComponent) (loaded)
  │       └── foo (PythonScriptComponent) (loaded)
  ├── multiple_component_instances_defs_py (loaded)
  │   └── component.py (loaded)
  │       ├── bar (PythonScriptComponent) (loaded)
  │       └── foo (PythonScriptComponent) (loaded)
  ├── multiple_definitions_calls (loaded)
  │   └── defs.py (loaded)
  ├── script_python_decl (loaded)
  │   ├── component.py (loaded)
  │   │   └── load (PythonScriptComponent) (loaded)
  │   └── cool_script.py (loaded)
  ├── scripts (loaded)
  │   ├── defs.yaml[0] (PythonScriptComponent) (loaded)
  │   ├── defs.yaml[1] (PythonScriptComponent) (loaded)
  │   └── defs.yaml[2] (PythonScriptComponent) (loaded)
  └── triple_dash_scripts (loaded)
      ├── defs.yaml[0] (PythonScriptComponent) (loaded)
      ├── defs.yaml[1] (PythonScriptComponent) (loaded)
      └── defs.yaml[2] (PythonScriptComponent) (loaded)
  '''
# ---
# name: test_load_from_path.2
  '''
  ├── multiple_component_instances (built)
  │   └── component.py (built)
  │       ├── bar (PythonScriptComponent) (built)
  │       └── foo (PythonScriptComponent) (built)
  ├── multiple_component_instances_defs_py (built)
  │   └── component.py (built)
  │       ├── bar (PythonScriptComponent) (built)
  │       └── foo (PythonScriptComponent) (built)
  ├── multiple_definitions_calls (built)
  │   └── defs.py (built)
  ├── script_python_decl (built)
  │   ├── component.py (built)
  │   │   └── load (PythonScriptComponent) (built)
  │   └── cool_script.py (built)
  ├── scripts (built)
  │   ├── defs.yaml[0] (PythonScriptComponent) (built)
  │   ├── defs.yaml[1] (PythonScriptComponent) (built)
  │   └── defs.yaml[2] (PythonScriptComponent) (built)
  └── triple_dash_scripts (built)
      ├── defs.yaml[0] (PythonScriptComponent) (built)
      ├── defs.yaml[1] (PythonScriptComponent) (built)
      └── defs.yaml[2] (PythonScriptComponent) (built)
  '''
# ---
